<a href="https://colab.research.google.com/github/ememdeee/Python-Artificial-Intelligence-Stuff/blob/main/Market_Based_Analysis_(MBA).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd

In [ ]:
data=pd.read_csv('online_retail_II.csv', encoding="ISO-8859-1")

Invoice StockCode  ... Customer ID         Country
0       536365    85123A  ...     17850.0  United Kingdom
1       536365     71053  ...     17850.0  United Kingdom
2       536365    84406B  ...     17850.0  United Kingdom
3       536365    84029G  ...     17850.0  United Kingdom
4       536365    84029E  ...     17850.0  United Kingdom
...        ...       ...  ...         ...             ...
148509  549241     22383  ...     15719.0  United Kingdom
148510  549241     22628  ...     15719.0  United Kingdom
148511  549241     20725  ...     15719.0  United Kingdom
148512  549241     20723  ...     15719.0  United Kingdom
148513  549241     22355  ...         NaN             NaN

[148514 rows x 8 columns]

In [ ]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 148514 entries, 0 to 148513
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   Invoice      148514 non-null  object 
 1   StockCode    148514 non-null  object 
 2   Description  147948 non-null  object 
 3   Quantity     148513 non-null  float64
 4   InvoiceDate  148513 non-null  object 
 5   Price        148513 non-null  float64
 6   Customer ID  102235 non-null  float64
 7   Country      148513 non-null  object 
dtypes: float64(3), object(5)
memory usage: 9.1+ MB


In [ ]:
data=data.dropna()

In [ ]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 102235 entries, 0 to 148512
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   Invoice      102235 non-null  object 
 1   StockCode    102235 non-null  object 
 2   Description  102235 non-null  object 
 3   Quantity     102235 non-null  float64
 4   InvoiceDate  102235 non-null  object 
 5   Price        102235 non-null  float64
 6   Customer ID  102235 non-null  float64
 7   Country      102235 non-null  object 
dtypes: float64(3), object(5)
memory usage: 7.0+ MB


In [ ]:
data_plus= data [data['Quantity']>=0]
data_plus.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 99656 entries, 0 to 148512
Data columns (total 8 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Invoice      99656 non-null  object 
 1   StockCode    99656 non-null  object 
 2   Description  99656 non-null  object 
 3   Quantity     99656 non-null  float64
 4   InvoiceDate  99656 non-null  object 
 5   Price        99656 non-null  float64
 6   Customer ID  99656 non-null  float64
 7   Country      99656 non-null  object 
dtypes: float64(3), object(5)
memory usage: 6.8+ MB


In [ ]:
basket_plus= (data_plus[data_plus['Country']=="United Kingdom"].groupby(['Invoice','Description'])['Quantity']
              .sum().unstack().reset_index().fillna(0).set_index('Invoice'))
basket_plus

Description   4 PURPLE FLOCK DINNER CANDLES  ...  ZINC WILLIE WINKIE  CANDLE STICK
Invoice                                      ...                                  
536365                                  0.0  ...                               0.0
536366                                  0.0  ...                               0.0
536367                                  0.0  ...                               0.0
536368                                  0.0  ...                               0.0
536369                                  0.0  ...                               0.0
...                                     ...  ...                               ...
549237                                  0.0  ...                               0.0
549238                                  0.0  ...                               0.0
549239                                  0.0  ...                               0.0
549240                                  0.0  ...                               0.0
549241                                  0.0  ...                               0.0

[4460 rows x 2907 columns]

In [ ]:
def encode_units(x):
  if x <= 0:
    return 0
  if x >= 1:
    return 1

basket_encode_plus = basket_plus.applymap(encode_units)
basket_encode_plus

Description   4 PURPLE FLOCK DINNER CANDLES  ...  ZINC WILLIE WINKIE  CANDLE STICK
Invoice                                      ...                                  
536365                                    0  ...                                 0
536366                                    0  ...                                 0
536367                                    0  ...                                 0
536368                                    0  ...                                 0
536369                                    0  ...                                 0
...                                     ...  ...                               ...
549237                                    0  ...                                 0
549238                                    0  ...                                 0
549239                                    0  ...                                 0
549240                                    0  ...                                 0
549241                                    0  ...                                 0

[4460 rows x 2907 columns]

the next step is to filter out the transactions that is bought more than 1 item. Here’s how I did it.

In [ ]:
basket_filter_plus= basket_encode_plus[(basket_encode_plus > 0).sum(axis=1) >= 2]
basket_filter_plus

Description   4 PURPLE FLOCK DINNER CANDLES  ...  ZINC WILLIE WINKIE  CANDLE STICK
Invoice                                      ...                                  
536365                                    0  ...                                 0
536366                                    0  ...                                 0
536367                                    0  ...                                 0
536368                                    0  ...                                 0
536372                                    0  ...                                 0
...                                     ...  ...                               ...
549237                                    0  ...                                 0
549238                                    0  ...                                 0
549239                                    0  ...                                 0
549240                                    0  ...                                 0
549241                                    0  ...                                 0

[4106 rows x 2907 columns]

In [ ]:
pip install mlxtend

In [ ]:
from mlxtend.frequent_patterns import apriori
from mlxtend.frequent_patterns import association_rules

In [ ]:
frequent_itemsets_plus = apriori(basket_filter_plus, min_support=0.03,
                                 use_colnames=True).sort_values('support', ascending = False).reset_index(drop=True)

frequent_itemsets_plus['length'] = frequent_itemsets_plus['itemsets'].apply(lambda x:len(x))

frequent_itemsets_plus

support                              itemsets  length
0    0.159279  (WHITE HANGING HEART T-LIGHT HOLDER)       1
1    0.117633            (REGENCY CAKESTAND 3 TIER)       1
2    0.101072   (SET OF 3 CAKE TINS PANTRY DESIGN )       1
3    0.083293             (JUMBO BAG RED RETROSPOT)       1
4    0.082075       (ASSORTED COLOUR BIRD ORNAMENT)       1
..        ...                                   ...     ...
116  0.030443       (SET OF 20 KIDS COOKIE CUTTERS)       1
117  0.030443         (CREAM SWEETHEART MINI CHEST)       1
118  0.030443                (HOT BATHS METAL SIGN)       1
119  0.030443            (PAPER BUNTING WHITE LACE)       1
120  0.030200                  (VINTAGE SNAP CARDS)       1

[121 rows x 3 columns]

Finding The Association Between Frequently Bought Items

In [ ]:
association_rules(frequent_itemsets_plus,metric='lift',
                  min_threshold = 1).sort_values ('lift', ascending=False).reset_index(drop=True)

antecedents  ... conviction
0      (GREEN REGENCY TEACUP AND SAUCER)  ...   4.725363
1     (ROSES REGENCY TEACUP AND SAUCER )  ...   3.621552
2           (ALARM CLOCK BAKELIKE GREEN)  ...   3.081405
3            (ALARM CLOCK BAKELIKE RED )  ...   3.358926
4          (WOODEN FRAME ANTIQUE WHITE )  ...   2.339995
5    (WOODEN PICTURE FRAME WHITE FINISH)  ...   2.764044
6                (HEART OF WICKER SMALL)  ...   1.911594
7                (HEART OF WICKER LARGE)  ...   2.255441
8    (SET OF 3 CAKE TINS PANTRY DESIGN )  ...   1.642091
9    (SET OF 6 SPICE TINS PANTRY DESIGN)  ...   2.397142
10              (JAM MAKING SET PRINTED)  ...   1.655792
11            (JAM MAKING SET WITH JARS)  ...   1.663741
12  (WHITE HANGING HEART T-LIGHT HOLDER)  ...   1.253796
13    (RED HANGING HEART T-LIGHT HOLDER)  ...   2.597612

[14 rows x 9 columns]

The higher the lift value, the higher the association between the items willl. If the lift value is more than 1, it is enough for us to say that those two items are associated each other.

Dari data pertama dapat kita lihat padada support values (GREEN REGENCY TEACUP AND SAUCER) dan (ROSES REGENCY TEACUP AND SAUCER )	adalah 0.036 yang berarti ada 3.6% pembelian dimana 2 item itu terjual bersamaan.

Dari confidence, kami bahkan dapat mengekstrak lebih banyak informasi. Ingat bahwa nilai confidence dipengaruhi oleh antecedent dan consequent.

if the antecedent is higher than the consequent, then the rule that will be applied is rule number 1 (not number 2). vice versa. In this case, the antecedent value is higher than the consequent value. It means we will apply rule number 1 which is 𝐺𝑅𝐸𝐸𝑁 𝑅𝐸𝐺𝐸𝑁𝐶𝑌 𝑇𝐸𝐴𝐶𝑈𝑃 𝐴𝑁𝐷 𝑆𝐴𝑈𝐶𝐸𝑅 → 𝑅𝑂𝑆𝐸𝑆 𝑅𝐸𝐺𝐸𝑁𝐶𝑌 𝑇𝐸𝐴𝐶𝑈𝑃 𝐴𝑁𝐷 𝑆𝐴𝑈𝐶𝐸. In a more detail explanation, it means that a customer will tends to bought Roses Regency Teacup and Saucer AFTER they bought Green Regency Teacup And Saucer. Not in the other way around.

Dari hasil diatas we could generates several business insights as follows :
1.Item Placements. We could put ROSES REGENCY TEACUP AND SAUCER and GREEN REGENCY TEACUP AND SAUCER in a closer place, maybe in a same shelf or any other closer place.
2.Products Bundling. We could put ROSES REGENCY TEACUP AND SAUCER and GREEN REGENCY TEACUP AND SAUCER as a single bundle of product with a lower price compare to each price combined. This way will attract more sales and generates more income.
3.Customer Recommendation and Discounts. We could put Roses Regency Teacup and Saucer in the cashier, so that every time a customer bought Green Regency Teacup and Saucer, we could offer and recommend them to buy Roses Regency Teacup and Saucer with a lower price.